### PR TITLE
Improve relation center calculation

### PR DESCRIPTION
**Pull request by alexey-zakharenkov**
Solves #33 
The problem was caused by circular relation references. By mistake, the stop_area relation 7588528 in OSM contained the stop_area_group relation 7588561 as a member, and it was a member of that stop_area_group itself. The calculate_centers() procedure failed to calculate center for such relations since it required all children to have calculated centers. A [stop_area] relation without a center could not be assigned to any city, so the "stop is not connected to a station" error appeared.

After the PR, if a relation has a relation member the center of which cannot be calculated, it is simply ignored and the center is calculated based on other members.

### CODE REVIEW

File: process_subways.py [hunk 1]
```diff
@@ -71,114 +72,152 @@ def slugify(name):
...
     for el in elements:
         if el["type"] == "node":
             nodes[el["id"]] = (el["lat"], el["lon"])
-        elif el["type"] == "way":
-            if "nodes" in el:
-                calculate_way_center(el)
+        elif el["type"] == "way" and "nodes" in el:
+            # If elements have been queried via overpass-api with
+            # 'out center;' clause then ways already have 'center' attribute
+            if "center" in el:
```
Comment (AMStrakhov) [inline process_subways.py:170]: NIT: MB, put inside `calculate_way_center`?

And then:
```python
            if center := calculate_way_center(el, nodes):
                ways[el["id"]] = center
```

File: process_subways.py [hunk 2]
```diff
@@ -71,114 +72,152 @@ def slugify(name):
...
+            else relation_centers
+        )
+        if m_id in member_container:
+            center[0] += member_container[m_id][0]
+            center[1] += member_container[m_id][1]
+            count += 1
+    if count == 0:
+        return None
+    element["center"] = {"lat": center[0] / count, "lon": center[1] / count}
+    return (element["center"]["lat"], element["center"]["lon"])
```
Comment (AMStrakhov) [inline process_subways.py:149]:
```suggestion
    return element["center"]["lat"], element["center"]["lon"]
```

File: tests/test_center_calculation.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    """Test center calculation. Test data [should] contain among others
+    the following edge cases:
+      - an empty relation. It's element should not obtain "center" key.
+      - relation as member of relation, the child relation following the parent
+        in the OSM XML file.
+      - relation with incomplete members (broken references).
+      - relations with cyclic references.
+    """
+
+    OSM_DATA = "tests/assets/kuntsevskaya_transfer.osm"
```
Comment (AMStrakhov) [inline tests/test_center_calculation.py:18]: Relative paths will make it easier to run tests with PyCharm runner
```python
    ASSETS_PATH = Path(__file__).resolve().parent / "assets"
    OSM_DATA = str(ASSETS_PATH / "kuntsevskaya_transfer.osm")
    CORRECT_CENTERS = ASSETS_PATH / "kuntsevskaya_centers.json"
```

File: process_subways.py [hunk 4]
```diff
@@ -71,114 +72,152 @@ def slugify(name):
...
+            and way_nodes[0] == way_nodes[-1]
+        ):
+            break
+        center[0] += node_centers[nd][0]
+        center[1] += node_centers[nd][1]
+        count += 1
+    if count == 0:
+        return None
+    element["center"] = {"lat": center[0] / count, "lon": center[1] / count}
+    return (element["center"]["lat"], element["center"]["lon"])
```
Comment (AMStrakhov) [inline process_subways.py:103]:
```suggestion
    return element["center"]["lat"], element["center"]["lon"]
```

File: process_subways.py [hunk 5]
```diff
@@ -71,114 +72,152 @@ def slugify(name):
...
-                center[0] += member_container[m["ref"]][0]
-                center[1] += member_container[m["ref"]][1]
-                count += 1
-        if count == 0:
-            empty_relations.add(el["id"])
-        else:
-            el["center"] = {"lat": center[0] / count, "lon": center[1] / count}
-            relations[el["id"]] = (el["center"]["lat"], el["center"]["lon"])
-        return True
+    nodes: Dict[int, Tuple[float, float]] = {}  # id => (lat, lon)
```
Comment (AMStrakhov) [inline process_subways.py:157]: You can create an alias for `Tuple[float, float]` like `Pin = Tuple[float, float]`, and use it instead:
```python
Pin = Tuple[float, float]
nodes: Dict[int, Pin] = {} # id => (lat, lon)
```

File: process_subways.py (hunk 1 above)
Comment (alexey-zakharenkov) [reply process_subways.py]: +

File: process_subways.py (hunk 2 above)
Comment (alexey-zakharenkov) [reply process_subways.py]: \+

File: tests/test_center_calculation.py (hunk 3 above)
Comment (alexey-zakharenkov) [reply tests/test_center_calculation.py]: \+

File: process_subways.py (hunk 4 above)
Comment (alexey-zakharenkov) [reply process_subways.py]: \+

File: process_subways.py (hunk 5 above)
Comment (alexey-zakharenkov) [reply process_subways.py]: \+

**Review by AMStrakhov — APPROVED** [review]

**Review by KarinaRomanVK — APPROVED** [review]
